pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -413,21 +416,63 @@ private PinotResourceManagerResponse uploadSegment(File indexDir, SegmentMetadat
     }
 
     String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(segmentMetadata.getTableName());
-    OfflineSegmentZKMetadata offlineSegmentZKMetadata = _pinotHelixResourceManager.getOfflineSegmentZkMetadata(
-        offlineTableName, segmentName);
+    ZNRecord znRecord = _pinotHelixResourceManager.getZnRecord(offlineTableName, segmentName);
 
-    if (offlineSegmentZKMetadata == null) {
+    if (znRecord == null) {
       // Creating new segment, not refreshing
       downloadUrl = moveSegmentToPermanentDirectory(provider, tableName, segmentName, tempTarredSegmentFile);
       _pinotHelixResourceManager.addNewSegment(segmentMetadata, downloadUrl);
       response = PinotResourceManagerResponse.SUCCESS_RESPONSE;
       return response;
     }
+
+    OfflineSegmentZKMetadata offlineSegmentZKMetadata = new OfflineSegmentZKMetadata(znRecord);
+
+    long minionExpectedCrc = -1L;
+    boolean isMinionConfigured = offlineTableConfig.getTaskConfig() != null;
+    boolean minionHeaderPresent = headers.getHeaderString(HttpHeaders.USER_AGENT).contains(CommonConstants.Minion.MINION_HEADER);
+    if (isMinionConfigured) {
+      // Minion configured use case so that we write to zk sparingly
+      String segmentPushStatus = offlineSegmentZKMetadata.getSegmentPushStatus();
+      SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd'T'HH:mm:ss.SSS");
+      Date dt = new Date();
+      String currentHumanReadableTime = sdf.format(dt);
+      offlineSegmentZKMetadata.setSegmentPushStatus("COMMITTING-" + currentHumanReadableTime);
+
+      if (segmentPushStatus != null) {
+        if (tooLate(currentHumanReadableTime, sdf)) {
+          LOGGER.warn("Segment was not properly committed, replacing segment {} for table {}", segmentName, tableName);
```

Comment:
Bump a metric here. We can raise an alert if this condition hits. You should log it as an ERROR